### PR TITLE
Refactor Searcher.js using hooks

### PR DESCRIPTION
**Pull request by laka3000**
#### What does this PR do?
Refactor `src/js/components/DataTable/Searcher.js` using hooks

#### Where should the reviewer start?
`src/js/components/DataTable/Searcher.js`

#### What testing has been done on this PR?
Running `yarn storybook` and checking its behavior

#### How should this be manually tested?
Run `yarn storybook`

#### Any background context you want to provide?
No

#### What are the relevant issues?
https://github.com/grommet/grommet/issues/3394

#### Screenshots (if appropriate)
--

#### Do the grommet docs need to be updated?
No

#### Should this PR be mentioned in the release notes?
No

#### Is this change backwards compatible or is it a breaking change?
No

### CODE REVIEW

File: src/js/components/DataTable/__tests__/__snapshots__/DataTable-test.js.snap [hunk 1]
```diff
@@ -5128,118 +5128,83 @@ exports[`DataTable search 2`] = `
   -webkit-flex-direction: column;
   -ms-flex-direction: column;
   flex-direction: column;
-  -webkit-flex: 0 0 auto;
-  -ms-flex: 0 0 auto;
-  flex: 0 0 auto;
+  -webkit-flex: 1 1;
```
Comment (ShimiSun) [inline src/js/components/DataTable/__tests__/__snapshots__/DataTable-test.js.snap]: These changes points out a regression on how `flex` was calculated, the snapshots files shouldn't change for the refactoring effort (Clean snapshots are an indication of an intact component)

Comment (laka3000) [reply src/js/components/DataTable/__tests__/__snapshots__/DataTable-test.js.snap]: will look into it

Comment (ShimiSun) [reply src/js/components/DataTable/__tests__/__snapshots__/DataTable-test.js.snap]: Thanks buddy!

Comment (laka3000) [reply src/js/components/DataTable/__tests__/__snapshots__/DataTable-test.js.snap]: I believe the snapshots aren't correct in the master branch, here's what I've found out:

Here's what the search looks like without an input value:

image

But, with a value, it looks like this:

image


Now, if you take a look at the specs, there are two snapshots being taken, one before the input, and one with an input (`src/js/components/DataTable/__tests__/DataTable-test.js:107`):

```
    expect(container.firstChild).toMatchSnapshot();
    fireEvent.click(container.querySelector('[aria-label="focus-search-a"]'));
    const searchInput = container.querySelector('[name="search-a"]');
    >> expect(document.activeElement).toBe(searchInput);
    fireEvent.change(searchInput, {
      target: { value: '[' },
    });
    >> expect(container.firstChild).toMatchSnapshot();
```

However, in the master branch, both snapshots haves the same values for the `flex`attributes (`src/js/components/DataTable/__tests__/__snapshots__/DataTable-test.js.snap:4867 and :5131`):

```
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
```

However, with the refactor, the snapshots are reflecting the visual changes the component goes through.

Comment (ShimiSun) [reply src/js/components/DataTable/__tests__/__snapshots__/DataTable-test.js.snap]: Interesting, thank you for taking a look at that.
I agree the snapshots should look different after using the search input field.
@ericsoderberghp I'd like your look at this snapshot as well.

**Comment by ShimiSun** [comment]
I had to resolve some conflicts that came from https://github.com/grommet/grommet/pull/3533, @ericsoderberghp I'll appreciate your second pair of eyes reviewing this PR, so we can move forward with it.

**Comment by ShimiSun** [comment]
Thank you for your contribution!! 🎉

**Pull request merged**